libraries/AP_HAL/FS.h

```diff
@@ -0,0 +1,58 @@
+#pragma once
+
+#include "AP_HAL_Namespace.h"
+
+#include <stdint.h>
+
+class AP_HAL::FS
+{
+public:
+    // returns a null-terminated string describing the last error that
+    // occured when using the library
+    virtual void last_errorstring(char *ret, uint8_t retlen) = 0;
+
+    virtual int open_for_read(const char *path) = 0;
+    virtual int open_for_write(const char *path) = 0;
+    virtual void close(int fd) = 0;
+
+    virtual int32_t seek_relative(int fd, int32_t off) = 0;
+    virtual int32_t seek_absolute(int fd, uint32_t off) = 0;
+
+    virtual int32_t read(int fd, uint8_t *buffer, uint32_t len) = 0;
```

Comment:
`len` should be `size_t` to avoid later casting shenanigans/issues.

Reply:
I've tried to avoid use of `ssize_t` and `size_t` and the like in these interfaces.  Lots of examples of that in here.

We should probably discuss whether my choice was a good one.  The basic idea was that this was an abstraction and shouldn't rely on the presence of ssize_t etc etc which are a posixy thing.